### PR TITLE
[ICDS Dashboard] Added app version and commcare version to system usage report

### PR DESCRIPTION
**Pull request by challabeehyv**
https://app.asana.com/0/1177124110930377/1174007365416122/f

##### SUMMARY
Added app version and commcare version to system usage report

##### FEATURE FLAG
yes- Enable access to pre-release features in the ICDS Dashboard reports

### CODE REVIEW

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py [hunk 1]
```diff
@@ -429,6 +429,31 @@ def updates(self):
...
+        FROM (
+            SELECT distinct 
+                awc_id,
+                supervisor_id,
+                %(start_date)s::DATE AS month,
+                LAST_VALUE(app_version) over w as app_version,
+                LAST_VALUE(commcare_version) over w as commcare_version
+            FROM "{usage_table}" 
+            WHERE month = %(start_date)s WINDOW w as (
+                PARTITION BY awc_id, supervisor_id ORDER BY
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:445]: is it possible to apply windows in the existing query above and let the `sum`s happen as they are happening right now and just find last_value for these two?  If that would keep the behavior of the existing query same then we should try that if that is not increasing the cost too much.

you can put that updated query here in the comment if above is true then I can try doing the benchmark before you doing the code change

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Left a few comments

File: custom/icds_reports/sqldata/exports/system_usage.py [hunk 2]
```diff
@@ -110,6 +110,18 @@ def columns(self):
                     format_fn=lambda x: (x or 0) if self.loc_level < 5 else "Not applicable at AWC level",
                     slug='num_supervisor_launched')
                 )
+            # adding version fields
+            if self.loc_level > 4 and self.config['month'] >= datetime.date(2020, 5, 1):
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/system_usage.py:114]: they are under feature flag only

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py (hunk 1 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:445]: SELECT
            awc_id,
            supervisor_id,
            month,
            sum(pse) AS usage_num_pse,
            sum(gmp) AS usage_num_gmp,
            sum(thr) AS usage_num_thr,
            sum(add_household) AS usage_num_hh_reg,
            sum(add_person) AS usage_num_add_person,
            sum(add_pregnancy) AS usage_num_add_pregnancy,
            sum(home_visit) AS usage_num_home_visit,
            sum(bp_tri1) AS usage_num_bp_tri1,
            sum(bp_tri2) AS usage_num_bp_tri2,
            sum(bp_tri3) AS usage_num_bp_tri3,
            sum(pnc) AS usage_num_pnc,
            sum(ebf) AS usage_num_ebf,
            sum(cf) AS usage_num_cf,
            sum(delivery) AS usage_num_delivery,
            LAST_VALUE(app_version) over w as app_version,
            LAST_VALUE(commcare_version) over w as commcare_version,
            CASE WHEN (
                sum(due_list_ccs) + sum(due_list_child) + sum(pse) + sum(gmp) + sum(thr)
                + sum(home_visit) + sum(add_pregnancy) + sum(add_household)
            ) >= 15 THEN 1 ELSE 0 END AS usage_awc_num_active,
            sum(due_list_ccs) AS usage_num_due_list_ccs,
            sum(due_list_child) AS usage_num_due_list_child_health
        FROM "{usage_table}"
        WHERE month = %(start_date)s GROUP BY awc_id, month, supervisor_id, app_version, commcare_version, form_date WINDOW w as (
                PARTITION BY awc_id, supervisor_id ORDER BY 
                form_date RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
            )

File: custom/icds_reports/ucr/data_sources/usage_forms.json [hunk 3]
```diff
@@ -964,6 +964,40 @@
           ]
         }
       },
+      {
+        "type": "expression",
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/ucr/data_sources/usage_forms.json:968]: added

File: custom/icds_reports/sqldata/exports/system_usage.py [hunk 4]
```diff
@@ -110,6 +110,18 @@ def columns(self):
...
+                agg_columns.append(DatabaseColumn(
+                    'AWW Application version',
+                     MaxColumn('app_version'), slug='app_version'
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/system_usage.py:117]: added them to grouping

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:428]: I think this query will become very complex. I thnk lets go ahead with a separate query. Also, we don't need to apply the month filter on quering on usage ucr. we just have to pull the lastest form

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Hey, I think all is fine, we though should go ahead with a separate query.

**Comment by challabeehyv** [comment]
> Hey, I think all is fine, we though should go ahead with a separate query.

okay..reverting

**Review by Rohit25negi — COMMENTED** [review]
Logically it looks good. I just checked the cost of the query,
It looks very huge and would increase the agg time by more than 15 minutes.
please give me some time to think to achieve the same thing with lower cost:

cost of query written in the PR:

```
 HashAggregate  (cost=0.00..0.00 rows=0 width=0)
   Group Key: remote_scan.awc_id, remote_scan.supervisor_id, remote_scan.month, remote_scan.app_version, remote_scan.commcare_version
   ->  Custom Scan (Citus Real-Time)  (cost=0.00..0.00 rows=0 width=0)
         Task Count: 64
         Tasks Shown: One of 64
         ->  Task
               Node: host=100.71.184.232 port=6432 dbname=icds_ucr
               ->  Unique  (cost=11219570.13..11531162.29 rows=2590355 width=82)
                     ->  Sort  (cost=11219570.13..11271502.15 rows=20772811 width=82)
                           Sort Key: "ucr_icds-cas_static-usage_forms_92fbe2aa".awc_id, "ucr_icds-cas_static-usage_forms_92fbe2aa".supervisor_id, (last_value("ucr_icds-cas_static-usage_forms_92fbe2aa".add_person) OVER (?)), (last_value("ucr_icds-cas_static-usage_forms_92fbe2aa".add_household) OVER (?)), "ucr_icds-cas_static-usage_forms_92fbe2aa".form_date
                           ->  WindowAgg  (cost=5846669.54..6365989.82 rows=20772811 width=82)
                                 ->  Sort  (cost=5846669.54..5898601.57 rows=20772811 width=78)
                                       Sort Key: "ucr_icds-cas_static-usage_forms_92fbe2aa".awc_id, "ucr_icds-cas_static-usage_forms_92fbe2aa".supervisor_id, "ucr_icds-cas_static-usage_forms_92fbe2aa".form_date
                                       ->  Seq Scan on "ucr_icds-cas_static-usage_forms_92fbe2aa_104250" "ucr_icds-cas_static-usage_forms_92fbe2aa"  (cost=0.00..1159434.44 rows=20772811 width=78)
                                             Filter: (form_date < '2020-02-01'::date)
(15 rows)



```

**Comment by challabeehyv** [comment]
closing this as  a differnet approach to https://github.com/dimagi/commcare-hq/pull/28091

**Pull request closed without merging**